daemon/indexing/apollo/index.js

```diff
@@ -167,23 +167,33 @@ const typeDefs = gql`
     rating: Int
   }
 
-  # TODO:
-  #  - Filtering definition needs more thinking. This is not flexible at all...
-  #  - Add location based filtering.
-  #  - Add fractional usage (e.g. availability) filtering.
+  enum ValueType {
+    STRING
+    FLOAT
+    DATE
+    ARRAY_STRING
+  }
+
+  enum FilterOperator {
+    EQUALS
+    CONTAINS #for array values where at least one must match E.g. list of categories 
+    GREATER
+    GREATER_OR_EQUAL
+    LESSER
+    LESSER_OR_EQUAL
+  }
+
+  # A generic listing filter
   input ListingFilter {
-    priceMin: inPrice
-    priceMax: inPrice
-    category: String
-    subCategory: String
-    locale: String
-    sellerAddress: String
-    buyerAddress: String
+    name: String!
+    value: String!
+    valueType: ValueType!
+    operator: FilterOperator!
   }
 
   # The "Query" type is the root of all GraphQL queries.
   type Query {
-    listings(searchQuery: String, page: Page, order: ListingOrder, filter: ListingFilter): ListingPage,
+    listings(searchQuery: String, page: Page, order: ListingOrder, filters: [ListingFilter!]): ListingPage,
```

Comment:
What does the syntax [ListingFilter!] mean exactly ? The filters argument should be optional, so shouldn't this just [ListingFilter] ?

Reply:
The syntax means that the filters argument is optional and that it must not contain null values. It can be an empty list, but not a list where any value is `null`.

Reply:
For what reason would you want to pass an empty filters list ?